include/vara/Feature/Constraint.h

```diff
@@ -25,55 +25,166 @@ class Constraint {
+      : BinaryConstraint(LeftOperand, RightOperand) {}
+
+  [[nodiscard]] std::string toString() const override;
+};
+
+class EquivalenceConstraint : public BinaryConstraint, public BoolschesConstraint {
+public:
+  EquivalenceConstraint(Constraint *LeftOperand, Constraint *RightOperand)
+      : BinaryConstraint(LeftOperand, RightOperand) {}
+
+  [[nodiscard]] std::string toString() const override;
+};
+
+class NegConstraint : public UnaryConstraint, public NumericConstraint {
+public:
+  NegConstraint(Constraint *Operand)
+      : UnaryConstraint(Operand) {}
+
+  [[nodiscard]] std::string toString() const override;
+};
+
+class AdditionConstraint : public BinaryConstraint, public NumericConstraint {
+public:
+  AdditionConstraint(Constraint *LeftOperand, Constraint *RightOperand)
+      : BinaryConstraint(LeftOperand, RightOperand) {}
+
+  [[nodiscard]] std::string toString() const override;
+};
+
+class SubtractionConstraint : public BinaryConstraint, public NumericConstraint {
+public:
+  SubtractionConstraint(Constraint *LeftOperand, Constraint *RightOperand)
+      : BinaryConstraint(LeftOperand, RightOperand) {}
+
+  [[nodiscard]] std::string toString() const override;
+};
+
+class MultiplicationConstraint : public BinaryConstraint, public NumericConstraint {
+public:
+  MultiplicationConstraint(Constraint *LeftOperand, Constraint *RightOperand)
+      : BinaryConstraint(LeftOperand, RightOperand) {}
+
+  [[nodiscard]] std::string toString() const override;
+};
+
+class DivisionConstraint : public BinaryConstraint, public NumericConstraint {
+public:
+  DivisionConstraint(Constraint *LeftOperand, Constraint *RightOperand)
+      : BinaryConstraint(LeftOperand, RightOperand) {}
+
+  [[nodiscard]] std::string toString() const override;
+};
+
+class LessConstraint : public BinaryConstraint, public NumericConstraint {
+public:
+  LessConstraint(Constraint *LeftOperand, Constraint *RightOperand)
+      : BinaryConstraint(LeftOperand, RightOperand) {}
+
+  [[nodiscard]] std::string toString() const override;
+};
+
+class GreaterConstraint : public BinaryConstraint, public NumericConstraint {
+public:
+  GreaterConstraint(Constraint *LeftOperand, Constraint *RightOperand)
+      : BinaryConstraint(LeftOperand, RightOperand) {}
+
+  [[nodiscard]] std::string toString() const override;
+};
+
+class LessEquelsConstraint : public BinaryConstraint, public NumericConstraint {
+public:
+  LessEquelsConstraint(Constraint *LeftOperand, Constraint *RightOperand)
+      : BinaryConstraint(LeftOperand, RightOperand) {}
+
+  [[nodiscard]] std::string toString() const override;
+};
+
+class GreaterEquelsConstraint : public BinaryConstraint, public NumericConstraint {
+public:
+  GreaterEquelsConstraint(Constraint *LeftOperand, Constraint *RightOperand)
+      : BinaryConstraint(LeftOperand, RightOperand) {}
+
+  [[nodiscard]] std::string toString() const override;
+};
+
 class PrimaryConstraint : public Constraint {
```

Comment:
What is the reasoning behind `Primary` as a constraint name? I'm not familiar with this.

Reply:
Primary represents the leave nodes in the constraint tree, i based this on the expression hierarchy from C; there is a primary expression either a constant, a literal and so on. In our case this is right now either a number or  a feature.